Gh 8112 query params in st.page_link

Asaurus1 opened the pull request:
## Describe your changes
Adds features for https://github.com/streamlit/streamlit/issues/8112

## Testing Plan
* Unit tests added for page switching
* Still need to see if there are any JS tests that need to be updated

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-wihuang commented on frontend/app/src/App.tsx:
```diff
@@ -1454,25 +1456,36 @@ export class App extends PureComponent<Props, State> {
       // click the "Rerun" button in the main menu. In this case, we
       // rerun the current page.
       pageScriptHash = currentPageScriptHash
+      // We also want to presever the existing query params _if_ they weren't specified
```
super nit: should this be `preserve`?

sfc-gh-wihuang commented on lib/streamlit/runtime/state/query_params_proxy.py:
```diff
@@ -205,3 +207,19 @@ def from_dict(self, params):
             return qp.from_dict(params)
+
+    @gather_metrics("query_params.to_string")
+    def to_string(self) -> str:
```
super nit: If this is internal, should probably be called `_to_string` and remove the `@gather_metrics`

sfc-gh-wihuang replied:
Is there a reason why this is added to the proxy and not the query_params class?

Asaurus1 replied:
It's in both...? Because we don't know if we're going to receive a proxy or the underlying object.

Technically users can use this to generat easy "permalinks" for people to copy, so it's ok in my mind to have it be private.

sfc-gh-wihuang replied:
Ah oops sorry misread that sorry 😨

sfc-gh-wihuang commented on frontend/app/src/App.tsx:
```diff
@@ -1742,6 +1762,19 @@ export class App extends PureComponent<Props, State> {
     return queryString.startsWith("?") ? queryString.substring(1) : queryString
   }
 
+  // queryParamsToString = (queryParams: Map<string, string[]>): string => {
```
super nit: could remove this

sfc-gh-wihuang commented on lib/streamlit/elements/widgets/button.py:
```diff
@@ -744,6 +768,20 @@ def _page_link(
             )
 
+        if query_params is not None:
+            if not hasattr(query_params, "get_all"):
```
Why is there this check out of curiosity? I assume it's because to check that we can use it to do `from_dict` but maybe could use a comment here why this check.

Asaurus1 replied:
If the `query_params` has a `get_all` method we assume (perhaps incorrectly) that it is a QueryParams object or some derivative. It's kind of a kludgy check, since we never actually _call_ `get_all` at all -- we call `to_string()` instead, but checking for `to_string()` seems like it would open us up to unintended behavior.

Perhaps this should be an `isinstance` check, and ONLY special-case objects that are `QueryParams` or `QueryParamsProxy`s?

sfc-gh-pchiu replied:
I'd lean towards using `isinstance` check. (I believe there's no other `QueryParams` derivatives other than `QueryParamsProxy`)

sfc-gh-wihuang commented:
This seems pretty good from first glance. Can we fix the github actions? 

Also, it seems like there are some tests that are breaking. Could we check those out?

Asaurus1 commented:
> This seems pretty good from first glance. Can we fix the github actions?
> 
> Also, it seems like there are some tests that are breaking. Could we check those out?

Yeah, this PR isn't ready yet, as it doesn't actually implement the fix. I am still stuck on how to make the JS frontend do what I want (could use some help on that, actually) but once that's fixed then of course we'll fix all the build failures.

sfc-gh-wihuang commented:
> > This seems pretty good from first glance. Can we fix the github actions?
> > Also, it seems like there are some tests that are breaking. Could we check those out?
> 
> Yeah, this PR isn't ready yet, as it doesn't actually implement the fix. I am still stuck on how to make the JS frontend do what I want (could use some help on that, actually) but once that's fixed then of course we'll fix all the build failures.

Ha yeah I was actually looking into the frontend too and I was also stuck on how to make the JS frontend work. What problem are u facing on the frontend? I was facing problems where the URL and backend state was not synced. I have checked out to your code so I will try to check it out and we can communicate back and forth.

Asaurus1 commented:
@sfc-gh-wihuang @sfc-gh-jcarroll can probably set you up with my Discord tag if you want to use that.

I'll see if I can put together a diagram or something of what I THINK the front-end is doing when you switch a page. I think the core issue is that the frontend clears the url query params when the page name changes every time it receives a "new session" (e.g. a new script runner) message. It DOESN'T get any information from that new script runner on what the backend thinks the query params currently _are_.

Asaurus1 commented:
@willhuang1997 ok so it "works" now (in the sense that, you can add query_params on the page switch). However, there are still some string joining issues that need to be cleaned up.

The bigger issue is that the way this is being done, we end up with multiple states getting added to browser history on the switch (one for the query_params change, and then a separate one for the actual page change). I am pretty sure that in order to avoid that we'd either need to 
A) do some hacky stuff in the python QueryParams object
B) do some hacky stuff on the frontend
C) rearchitect some of the way communication works between front/back when page changes happen to remove the need for the duplicate messages.

I am still putting together that diagram to try to better understand what all is involved in the communication path between the two sides.

sfc-gh-wihuang commented:
> @willhuang1997 ok so it "works" now (in the sense that, you can add query_params on the page switch). However, there are still some string joining issues that need to be cleaned up.
> 
> The bigger issue is that the way this is being done, we end up with multiple states getting added to browser history on the switch (one for the query_params change, and then a separate one for the actual page change). I am pretty sure that in order to avoid that we'd either need to A) do some hacky stuff in the python QueryParams object B) do some hacky stuff on the frontend C) rearchitect some of the way communication works between front/back when page changes happen to remove the need for the duplicate messages.
> 
> I am still putting together that diagram to try to better understand what all is involved in the communication path between the two sides.

Thanks for researching this! I added you on discord but not sure if you accepted my discord friend req.

Asaurus1 commented:
@sfc-gh-wihuang I don't think I ever got your discord, or perhaps it got lost, could you send it again?

sfc-gh-wihuang commented:
> @sfc-gh-wihuang I don't think I ever got your discord, or perhaps it got lost, could you send it again? 

My profile is williamhuang!

Asaurus1 commented:
@willhuang1997 I finally got your contact in discord. Working on diagraming out the updates we need to make and then I'll send them over or post them here.

@vdonato I think this will definitely touch codepaths which control how the App interacts with the browser window history, I want to avoid any regressions back to something like https://github.com/streamlit/streamlit/pull/6271#issuecomment-1465090690;, or affecting fragment behavior. I don't have a working set of changes yet, but once I do, would you do a buddy review as well since you worked on 6271?

vdonato commented:
> I don't have a working set of changes yet, but once I do, would you do a buddy review as well since you worked on 6271?

Hey @Asaurus1, sure! I'll try to give this PR a review once you have things working. Heads up that I'll likely be doing a bit less work on the open source maintainer side for a few months, though, so there's a chance another team member will have to pick this PR review up instead of me.

Asaurus1 commented:
@vdonato I have a "working" version that you can run by including the following new page in the `streamlit hello` set of pages and then running it.

```python
import streamlit as st
import time

st.write(st.query_params)

def _increment():
    for x in range(3):
        st.query_params["test"] = int(st.query_params.get("test", 0)) + 1
        time.sleep(1)

st.button("count to 3", key="foo", on_click=_increment)

if st.button("switch page", key="switch"):
    st.switch_page("Animation_Demo.py", query_params={"switch": "page"})

st.page_link("Hello.py", query_params={"boo": "far"})
```

```python
# in hello/streamlit_app.py
...
            st.Page(dir_path / "Mapping_Demo.py", icon=":material/public:"),
            st.Page(dir_path / "Dataframe_Demo.py", icon=":material/table:"),
            st.Page(dir_path / "Queryparams_Demo.py", icon=":material/question_mark:"),
        ]
    )
```

Things that are still broken and will need to be fixed:
* the back button behavior is weird (#9279 still occurring, it might even be worse now)
* There may be some places you get more than one entry in `window.history` for a single click
* My hair is gray and falling out...:P

Some open questions that I have which will need to be addressed before we can really say this change has been implemented in a "good" way:
* There seems to be two different paradigms in the codebase; some functions pull the queryString directly from `window.location.search`, while other functions use `App.getQueryParams()` which ultimately defaults to `App.state.queryParams` if that value exists, and otherwise falls back on `window.location.search`; I am worried that the two can get out of sync somehow, and I'm not really sure why we have the queryParams in `App.state` at all. If a maintainer could help explain this that would be great.
* There are a bunch of query_string-related messages in "HostConnectionManager" that seem to have no senders/receivers in the streamlit codebase. My assumption here is that this has to do with the integration into CommunityCloud. If that is the case then the changes I have made have probably royally screwed up assumptions about the ordering and contents of those messages. We should get someone from the CommunityCloud team to come and inspect this as well. If, on the other hand, they're just left over code that's no longer used, I'm happy to remove them.

Asaurus1 commented:
Browser navigation seems to work for the most part now, and I haven't been seeing instances of getting double-entries. 
Still hoping to get some clarifying information from the maintainers on the other questions above.

Asaurus1 commented:
For reviewers, here is more or less what has changed
![image](https://github.com/user-attachments/assets/0b5e37ff-d0ec-4510-b6b1-daac661c7c70)
![image](https://github.com/user-attachments/assets/2d1ab5b9-9c40-4692-94bb-6364988e87f5)

Asaurus1 commented:
@lukasmasuch due to the extensive nature of this update, I do want a maintainer to take a look and provide feedback on the implementation prior to me going about cleaning up all the tests that this has broken - I'd rather not rewrite the tests and then we decide to change the implementation after.

I did upgrade the typescript version and yarn packages in this branch because the existing version of TS that you guys were using had a bug where it didn't recognize the `URLSearchParams.size` property. This drove additional compilation and linting changes to other files, but no major side effects in terms of functionality. We can strip out that upgrade (removing the `.size` checks that drove it) if you guys already have a timeline for upgrading TS in the project somewhere, otherwise it seems like a good thing to keep to me.

lukasmasuch commented:
@Asaurus1 Unfortunately, I haven't had time to look closer into this feature. We discussed this internally and someone from the team will pick this up soon and do a proper review. It would be awesome if we can get it in for 1.39 which is still a couple of weeks away.

Asaurus1 commented:
@lukasmasuch I've reopened this PR and merged-forward to the latest library version -- fortunately there were minimal merge conflicts and the basic behavior changes still work like they did when I last tested it. I think last time we talked about this you mentioned wanting to get it in a November patch?

lukasmasuch commented:
@Asaurus1 sorry for the delay :( @sfc-gh-pchiu will take over reviewing this feature and I believe its still planned to get it merged until end of this month.

sfc-gh-pchiu commented on Makefile:
```diff
@@ -262,7 +262,7 @@ protobuf: check-protoc
 		yarn --silent pbjs \
 			../proto/streamlit/proto/*.proto \
-			--path=proto -t static-module --wrap es6 \
+			--path=../proto -t static-module --wrap es6 \
```
do we need this change?

Asaurus1 replied:
it's been a while since I made this commit on this branch -- and I didn't exactly leave myself the most clear commit message. But IIRC this was required with the upgrade to ES6. I'll do some testing and if it still works the old way I'll roll it back.

Asaurus1 replied:
The typescript upgrade was originally necessitated by not being able to do iteration through various types like `URLSearchParamsIterator` or `Set<string>` in some of the utility functions. 

![image](https://github.com/user-attachments/assets/fd210107-049d-4f39-876e-88925ad7b269)

I think it's viable to find uglier work arounds for these and move the TS upgrade to a [separate PR](https://github.com/Asaurus1/streamlit/tree/typescript-upgrade), but I do still think we should do the TS upgrade. TS v4.9.5 which Streamlit is currently on is about 2 years EOL.

sfc-gh-pchiu replied:
I checked with the team and we don't see any thing blocking the upgrade. So please feel free to publish a separate PR! We can get that merged prior to this one.

sfc-gh-pchiu commented on frontend/app/src/App.tsx:
```diff
@@ -1026,7 +1073,7 @@ export class App extends PureComponent<Props, State> {
     if (!fragmentIdsThisRun.length) {
       // This is a normal rerun, remove all the auto reruns intervals
-      this.state.autoReruns.forEach((value: NodeJS.Timer) => {
+      this.state.autoReruns.forEach((value: NodeJS.Timeout) => {
```
nit: let's move this to another PR

sfc-gh-pchiu commented on frontend/app/src/App.tsx:
```diff
@@ -1097,7 +1144,11 @@ export class App extends PureComponent<Props, State> {
+      this.handleSessionStatusChanged(
+        initialize.sessionStatus as SessionStatus
+      )
+    }
```
nit: let's move this to another PR

sfc-gh-pchiu commented on frontend/app/src/App.tsx:
```diff
@@ -1106,17 +1157,27 @@ export class App extends PureComponent<Props, State> {
 
     // do not cause a rerun when an anchor is clicked and we aren't changing pages
-    const hasAnchor = document.location.toString().includes("#")
+    const hasAnchor = window.location.toString().includes("#")
```
Do we need to change to using `window.location`? 
It seems like it's possible to get the correct `targetQueryString` from `document.location`, too.

Asaurus1 replied:
A little internet research indicates that `window.location` is the more supported reference to use for older browsers, although most modern browsers map them to the same object.

sfc-gh-pchiu replied:
fair point

sfc-gh-pchiu commented on frontend/app/src/App.tsx:
```diff
@@ -1455,11 +1516,23 @@ export class App extends PureComponent<Props, State> {
-      pageScriptHash
-    )
+    // clear non-embed query parameters within a page change
+    const queryParams = preserveEmbedQueryParams(queryString ?? "").toString()
```
I assume this is moved from `sendRerunBackMsg` -- this would now run, i.e. clearing the non-embed query params, if `pageScriptHash == currentPageScriptHash`, right?

sfc-gh-pchiu commented on frontend/app/src/components/AppView/styled-components.ts:
nit: let's move these improvements to another PR

Asaurus1 replied:
To be exact, these changes are due to the Typescript upgrade from v4 to v5. My recollection from earlier this year was that this upgrade was required in order to implement some of the features on this PR. After re-basing, though, I will go back and double check that it is still necessary -- if not, I am happy to move them to a separate PR.

sfc-gh-pchiu commented on frontend/app/src/components/Sidebar/styled-components.ts:
ditto

sfc-gh-pchiu commented on frontend/app/src/components/StreamlitDialog/DeployErrorDialogs/styled-components.tsx:
ditto

sfc-gh-pchiu commented on frontend/app/src/components/StreamlitDialog/styled-components.ts:
ditto

sfc-gh-pchiu commented on frontend/app/src/index.tsx:
ditto

sfc-gh-pchiu commented on lib/streamlit/elements/widgets/button.py:
```diff
@@ -838,6 +854,16 @@ def _page_link(
                     main_script_directory=main_script_directory,
                 )
 
+        if query_params is not None:
+            if not hasattr(query_params, "get_all"):
+                query_params_internal: QueryParams = QueryParams()
+                # turn off sending forward messages for THIS QueryParams
+                query_params_internal._disable_forward_msg = True
+                query_params_internal.from_dict(query_params)
+            else:
+                query_params_internal = cast(QueryParams, query_params)
```
As mentioned, do we need the `QueryParams` obj or could we generate the string directly?

Asaurus1 replied:
You _could_ generate the string correctly, however the `QueryParams` class has a number of protections and translation layers in place that make it safer, DRYer, and behaviorally consistent to just use the class to generate the string. For example, `QueryParams` does not allow the user to specify `embed` or any of the other embed-like keys, since those have special meaning. And `QueryParams` takes care of translating list-like objects in the incoming mapping into repeated keys in the query string.

sfc-gh-pchiu replied:
Thanks, this is well explained and I agree we should use it here.

sfc-gh-pchiu commented on proto/streamlit/proto/NewSession.proto:
```diff
@@ -69,6 +69,10 @@ message NewSession {
+
+  // Any query string that has been set by the backend for this session; should
+  // not include any embed= or embed_theme= keys.
+  string query_string = 12;
```
may have missed this -- are we using this field? Could we rely on the field in `PageLink.proto` to be consumed in `PageLink.tsx`?

Asaurus1 replied:
You are correct about `PageLink`, however you need this field for `switch_page()`, which spawns a new re-run on the backend. Including the `query_string` field in the `NewSession` is the best and (currently) only way for it to tell the front-end the query string has changed.

sfc-gh-pchiu reviewed (commented):
Hi @Asaurus1, thank you very much for the patience. I have some bandwidth this cycle to help look into this. 

The `page_link` command with `query_params` I repro'ed locally seems to work as expected. The design makes sense to me as well so we could go from here.  
```
# my test snippet

import streamlit as st

pages = []

def page_1():
    st.page_link(pages[1], query_params={"type": "bar"}, label="Go to page 2")
    st.write(f"Param: {st.query_params}")

def page_2():
    st.write(f"Param: {st.query_params}")

pages = [st.Page(page_1), st.Page(page_2)]
st.navigation(pages).run()
```

A few high-level comments I have:
* code improvements tangential to this feature -- It'd be awesome if they could be separated out to new PR(s), they could get more eyes from the team, joining the overarching effort to improve our quality.
* `App.tsx -> handlePageInfoChanged()` -- I may be missing something but I wonder if we need to change this event handler. The page info `msgProto` didn't get sent in my test and the window history stack looks correct.
* `button.py -> _page_link()` -- (maybe a naive point) I am curious about the purpose of instantiating a `QueryParams` object v.s. directly setting the `query_string` field for `PageLinkProto`. What'd be the downside if we do the latter?

Again, thanks so much for the effort! I may not get clarity on all the details in the first pass, so please feel free correct me as we go 🙏

Asaurus1 commented:
Thank you for picking up the review. I am back from traveling and will have some more time to work on this as well.

First step will be re-basing the branch on top of develop in order to pull in all the intermediary upgrades. That may take a day or two here depending on how many tests break. Once it's done I'll let you know.

> * `App.tsx -> handlePageInfoChanged()` -- I may be missing something but I wonder if we need to change this event handler. The page info `msgProto` didn't get sent in my test and the window history stack looks correct.

I am not sure I follow this one off the top of my head (marking with 👀 here so I remember to go do a little more digging). Would you be able to elaborate or rephrase the question?

> * `button.py -> _page_link()` -- (maybe a naive point) I am curious about the purpose of instantiating a `QueryParams` object v.s. directly setting the `query_string` field for `PageLinkProto`. What'd be the downside if we do the latter?

In short, behavioral consistency and DRY principles: see https://github.com/streamlit/streamlit/pull/9049#discussion_r1835988169

Commit:
GH-8112: add query_params to NewSession and PageLink protos

Commit:
GH-8112: update QueryParams class to allow "silent" instances for internal use without sending forward messages.

Commit:
GH-8112: add query_params argument to st.page_link

Commit:
GH-8112: update the queryString manipulator utility functions so that we have better control of information flow

Commit:
GH-8112: maybeUpdatePageUrl handles checking for changes in the query string vs app state, and handles ALL window history updates.

app tsx imports

Commit:
GH-8112: onPageChange accepts a query_string parameter; allow PageLink to forward it's configured query_string (from it's proto) to the page change call.

Commit:
GH-8112: Add query_params argument to st.switch_page and machinery to forward query string to the frontend on every new session

Commit:
GH-8112: AppNavigation strategies handle query string in NewSession messages arriving from the backend and update app state appropriately

Commit:
GH-8112: apply consistency to usage of window.location instead of document.location, and fix up some failing tests due to new usage.

Commit:
GH-8112: add e2e tests for st.page_link and st.switch_page

Asaurus1 commented:
@sfc-gh-pchiu ok, I put a bunch of work into rebasing all the work on this branch so far so that the individual commits represent incremental changes, which hopefully should help with review. I've also added some additional tests both for the new features, as well as in a few places where existing features I used in this change were found to be lacking testing.

The last commit is some e2e tests of the feature. This makes it quite easy to demo as well, just run

```bash
streamlit run e2e_playwright/multipage_apps_v2/mpa_v2_with_query_params.py
```

One thing I could use some help on -- as part of this process I seem to have introduced a flaky behavior for `st.switch_page` when V2 Navigation is in use. This can be seen in flakiness of the final two tests [here](https://github.com/streamlit/streamlit/pull/9049/files#diff-4718fabf55040cb0b1d1c2ca8bc13a7c825cce46f171a6a89a7bd54b7118c617R135-R164). You can see the same thing when running the demo script above -- occasionally using the `st.switch_page` feature results in the page switch showing up in the URL bar, but the new page is entirely devoid of elements.

I am convinced this is some race condition between the "clear-all-current-widgets-when-the-page-switches" action in `App.tsx` and the element messages coming from the backend for the new elements that are supposed to go on the new page. I have tried to narrow in on the problem by adding logging statements, but because it's a timing issue, logging seems to change the timing of things and I can't actually catch the behavior. 

I'm kind of hoping one of the SFC engineers might be able to lend some advice on how to tackle this.

Asaurus1 commented on frontend/app/src/App.tsx:
```diff
@@ -308,7 +308,7 @@ export class App extends PureComponent<Props, State> {
       sidebarChevronDownshift: 0,
       pageLinkBaseUrl: "",
-      queryParams: "",
+      queryParams: null,
```
Note: this change from `""` to `null` as a default value is so that we can more easily differentiate when the queryParams have been reset to empty vs being uninitialized at App startup. It makes the logic check in `App.getQueryString()` much easier to understand.

Asaurus1 commented on frontend/app/src/App.tsx:
```diff
@@ -921,37 +919,74 @@ export class App extends PureComponent<Props, State> {
 
-        window.history.pushState({}, "", pageUrl)
+      // If queryString was specified, send a host message
+      if (queryString) {
```
I am not 100% certain that this is a correct place to send a `SET_QUERY_PARAM` message to the host. This is partly because I don't fully understand what `hostCommunicationMgr` is all about -- I think it's something to do with Community Cloud, as it doesn't seem to have any effect when running the application locally.

My reason for being uncertain is this: `maybeUpdatePageUrl`, with the changes introduced in this PR, is _only_ ever called after `App.state.queryString` has been updated -- this is true by design, since the method _gets_ `requestedQueryString` from `App.getQueryString()` which in most cases, loads the value from `App.state`. 

There are [some places](https://github.com/streamlit/streamlit/pull/9049/files#diff-1b95d4e0dda1e5928f99c8303599bd811f140fdd40e028fcfca3077f802d6210R1492) where we also send this message _prior_ to updating `App.state.queryString` but [some places](https://github.com/streamlit/streamlit/pull/9049/files#diff-7f6abc01745f6d3e4805e1279acec8c843ba151febe680c576d1497a3fb5a077R105-R112) where we do not. I do not really understand which is correct. When does the host that is on the other side of this _expect_ to receive this message and what behavior does it induce when it receives one?

sfc-gh-pchiu replied:
I checked with the team and `hostCommunicationMgr` handles communication between a Streamlit app & its parent frame (so yes, Community Cloud relies on it and not for local apps.)

I would thinking that we'd update `App.state.queryString` when sending `SET_QUERY_PARAM` message to host. In the example where we do not, I wonder if you mean that there's no update to `App.state.appPages`?

Asaurus1 commented:
I have also been wondering if, just like we have a `MetricsManager` and a `WidgetStateManager` for the frontend `App` right now, perhaps this change is large enough that it would be cleaner to take all the functionality and pack it into a `UrlStateManager` or `QueryStringManager`?

Commit:
GH-8112: onHistoryChange should still call `onPageChange` when the query string changes, even if there is an anchor in the URL

Asaurus1 commented:
> One thing I could use some help on -- as part of this process I seem to have introduced a flaky behavior for `st.switch_page` when V2 Navigation is in use. This can be seen in flakiness of the final two tests [here](https://github.com/streamlit/streamlit/pull/9049/files#diff-4718fabf55040cb0b1d1c2ca8bc13a7c825cce46f171a6a89a7bd54b7118c617R135-R164). You can see the same thing when running the demo script above -- occasionally using the `st.switch_page` feature results in the page switch showing up in the URL bar, but the new page is entirely devoid of elements.

@sfc-gh-pchiu and @kmcgrady I've rebased onto develop and fixed up what I think are all of the test failures. 

I tried as best I could to get this "flaky test" I mentioned above to fail again, but after rebasing it seems to have stubbornly disappeared, and I'm unable to repro it any more. With that being the case, then, everything seems to be working and I would say this PR is feature-complete.

Asaurus1 commented:
Rebased to fix some breaking tests.

jrieke commented:
Hey @Asaurus1, sorry that nobody got back here. Eric has been pretty busy with another project, so he didn't have time to look into this yet. But the other project should wrap up within the next few weeks, so I hope we'll have a bit more time for this PR then.

Asaurus1 commented:
> Hey @Asaurus1, sorry that nobody got back here. Eric has been pretty busy with another project, so he didn't have time to look into this yet. But the other project should wrap up within the next few weeks, so I hope we'll have a bit more time for this PR then.

Rebased and pushed.

mickaelandrieu commented:
Do we have any workaround in the meantime, i'm trying to finish a basic CRUD app and I'm living a nightmare for no decent reason...

A basic code like :

> bots['update_link'] = "/update?bot_id=" + urllib.parse.quote(str(bots['bot_id']))

Used in an st.data_editor doesnt works.

sfc-gh-pchiu commented on lib/streamlit/elements/widgets/button.py:
```diff
@@ -646,6 +650,12 @@ def page_link(
             Whether to expand the link's width to fill its parent container.
             The default is ``True`` for page links in the sidebar and ``False``
             for those in the main app.
+        query_params : dict[str, str]
```
this would be `dict[str, str | list[str]]`?

sfc-gh-pchiu commented on lib/streamlit/commands/execution_control.py:
```diff
@@ -148,11 +148,14 @@ def rerun(  # type: ignore[misc]
-def switch_page(page: str | Path | StreamlitPage) -> NoReturn:  # type: ignore[misc]
+def switch_page(  # type: ignore[misc]
+    page: str | Path | StreamlitPage,
+    query_params: Mapping[str, str | Iterable[str]] | None = None,
```
We have official doc reviews for releases, but some preliminary/placeholder docs for the new param would be helpful. (could be the same as `page_link`)

sfc-gh-pchiu commented on lib/streamlit/commands/execution_control.py:
```diff
@@ -223,14 +226,20 @@ def switch_page(page: str | Path | StreamlitPage) -> NoReturn:  # type: ignore[m
     )
     # Force a yield point so the runner can do the rerun
-    st.empty()
+    ctx.session_state._yield_callback()
```
I guess this should be fine -- I understood it as achieving the same purpose w/o changing the UI?
@kmcgrady I see `_yield_callback()` was introduced in #7373 and wonder if you could help me validate if this and the 2 same changes in this file make sense.

sfc-gh-pchiu commented on lib/streamlit/commands/execution_control.py:
```diff
@@ -223,14 +226,20 @@ def switch_page(page: str | Path | StreamlitPage) -> NoReturn:  # type: ignore[m
-        qp.clear()
+    # but also allow the user to update the query_params explicitly.
+    if isinstance(query_params, QueryParams):
+        _params_to_send = query_params
+    else:
+        _params_to_send = QueryParams()
+        _params_to_send._disable_forward_msg = True
+        if query_params is not None:
+            _params_to_send.from_dict(query_params)
 
     ctx.script_requests.request_rerun(
         RerunData(
-            query_string=ctx.query_string,
+            query_string=_params_to_send.to_string(),
```
If `query_params is None`, we may still need to reset `ctx.session_state.query_params()`.
Also if there are embed qps in the session state, do we have a way to keep them in the new query string?

sfc-gh-pchiu commented on frontend/tsconfig.json:
```diff
@@ -3,6 +3,7 @@
     "allowJs": true,
     "allowSyntheticDefaultImports": true,
     "baseUrl": ".",
+    "downlevelIteration": true,
```
Do we still need this?

sfc-gh-pchiu commented on lib/streamlit/runtime/state/query_params.py:
```diff
@@ -182,6 +189,11 @@ def from_dict(
             # restore the original from before we made any changes.
             self._query_params = old_value
             raise
+        return self
```
Do we still need this change? It looks like we don't need the return value in other places.

sfc-gh-pchiu commented on lib/streamlit/runtime/state/query_params.py:
```diff
@@ -41,6 +43,7 @@ class QueryParams(MutableMapping[str, str]):
     """
 
     _query_params: dict[str, list[str] | str] = field(default_factory=dict)
+    _disable_forward_msg: bool = False
```
I've been re-thinking about this flag. Given that we want leverage the logic in `QueryParams` class without the need to send messages, I wonder if inheritance could be an alternative here.

The base class would have everything except for the forward message part (e.g. making `_send_query_param_msg` no-op)

```py
# class names tbd
class BaseQueryParams(MutableMapping[str, str]):
    ...
    def _send_query_param_msg(self) -> None:
        pass
    ...

class QueryParams(BaseQueryParams):
    def _send_query_param_msg(self) -> None:
        ctx = get_script_run_ctx()
        ...
```

sfc-gh-pchiu commented on lib/tests/streamlit/runtime/state/session_state_test.py:
```diff
@@ -890,7 +890,7 @@ def test_session_state_stats(self):
         # The expected size of the session state in bytes.
         # It composes of the session_state's fields.
-        expected_session_state_size_bytes = 3000
+        expected_session_state_size_bytes = 3065
```
This is for the `query_string` field I suppose. Is there a reason why the delta is 65 bytes?

sfc-gh-pchiu commented on lib/tests/streamlit/elements/page_link_test.py:
```diff
@@ -153,3 +157,19 @@ def test_st_page_with_url_path(self):
+        page = st.Page("foo.py", title="QP Test")
+        st.query_params["foo"] = ["bar", "baz"]
+        st.query_params["marvin"] = "paranoid"
+        st.page_link(page=page, query_params=st.query_params)
```
Could we also test when the input is not an instance of QueryParams/QueryParamsProxy?

sfc-gh-pchiu commented on frontend/app/src/App.tsx:
```diff
@@ -925,37 +923,74 @@ export class App extends PureComponent<Props, State> {
-  maybeUpdatePageUrl = (
-    mainPageName: string,
-    newPageName: string,
-    isViewingMainPage: boolean
```
Should we keep this param? I see that its value is derived from [comparing page script hashes](https://github.com/streamlit/streamlit/blob/develop/frontend/app/src/util/AppNavigation.ts#L88-L89) which may not be exactly comparing `mainPageName` and `newPageName`

sfc-gh-pchiu reviewed (commented):
Hi @Asaurus1, I have more time now to pick this back up. I left some comments and open questions to be discussed. 
Also, there are some changes to our develop that could impact this PR: 1) we've updated the typescript version to the latest 2) MPA strategy v1 is moving away soon
Please feel free to slack me (Eric Chiu) as well for any question!

Asaurus1 commented:
> Hi @Asaurus1, I have more time now to pick this back up. I left some comments and open questions to be discussed. Also, there are some changes to our develop that could impact this PR: 1) we've updated the typescript version to the latest 2) MPA strategy v1 is moving away soon Please feel free to slack me (Eric Chiu) as well for any question!

Hey there! Now it's my turn to apologize; life and work has been busy the past 2 month and I haven't had a chance to come back to this. Hoping middle of April will be a little different. I will definitely start by taking 1) and 2) into account.

mickaelandrieu commented:
You guys give your time for free to help the community, thank you very much !

sfc-gh-jrieke commented:
Hey @Asaurus1! No rush from our side at all but just wanted to ask if you still plan to finish up this PR or whether it would make more sense if we find an engineer from our team to finish it up? We're going through quarterly planning this/next week, that's why I'm curious. Either way is fine for us :)

jrieke commented:
Hey @Asaurus1! I'm going to close this PR for now because it's probably a bit outdated already. In case you plan to work further on this, feel free to reopen or make a new PR! If not, no worries, we're planning to get an engineer from our team to work on this sometime within the next ~6 months I hope.
